calibration_residual_kernals.py

```diff
@@ -0,0 +1,88 @@
+                else:
+                    freq_range = observation_frequencies[tune, frequency_indices[tune]]
+                    phases = np.angle(gain_matrix[:,frequency_indices[tune]])
+                    for pol in range(nof_pols):  #probably unecessary - could do with some matrix mult stuff
+                        #some binary logic
+                        stream_idx = int(str(tune)+str(pol),2)
+                        unwrapped_phases = np.unwrap(phases[stream_idx,:])
+                        phase_slope = np.polyfit(freq_range, unwrapped_phases,1)[0]
+                        residuals = unwrapped_phases - (phase_slope*freq_range)
+                        residual_delays[stream_idx] = (phase_slope / (2*np.pi)) * 1e9
+                        residual_phases[stream_idx,frequency_indices[tune]] = residuals % (2*np.pi)
+            
+            delay_residual_map[ant] = residual_delays
+            phase_residual_map[ant] = residual_phases
+
+        return delay_residual_map, phase_residual_map
+
+def calc_residuals_from_ifft(ant_to_gains, observation_frequencies):
+    """
+    Accept mapping of antenna to gains.
+    Returns ant to residual delay and ant to residual phase maps.
+
+    Args:
+        ant_to_gains : {<ant> : [[complex(gains_pol0_tune0)], [complex(gains_pol1_tune0)], [complex(gains_pol0_tune1)], [complex(gains_pol1_tune1)]], ...}
+        observation_frequencies : list of dimension (n_tunings, nchans)
+    Return:
+        delay_residual_map : {<ant> : [[residual_delay_pol0_tune0], [residual_delay_pol1_tune0], [residual_delay_pol0_tune1], [residual_delay_pol1_tune1]]}, ...} in nanoseconds
+        phase_residual_map : {<ant>_<tune_index> : [[residual_phase_pol0],[residual_phase_pol1]]}, ...} in radians
+    """
+    delay_residual_map = {}
+    phase_residual_map = {}
+    for ant, gain_matrix in ant_to_gains.items():
+        gain_matrix = np.array(gain_matrix, dtype = np.complex64)
+        nof_streams = gain_matrix.shape[0]
+        nof_tunings,nof_chan = observation_frequencies.shape
+        nof_pols = int(nof_streams/nof_tunings)
+        ifft_abs_matrix = np.abs(np.fft.ifft(gain_matrix, axis = 1))
+        phase_matrix = np.angle(gain_matrix)
+        max_idxs = np.argmax(ifft_abs_matrix,axis=1)
+        residual_delays = np.zeros(nof_streams,dtype=np.float64)
+        residual_phases = np.zeros(gain_matrix.shape,dtype=np.float64)
+
+        for tune in range(nof_tunings):
+            chan_width = observation_frequencies[tune,1] - observation_frequencies[tune,0]
+            tlags = np.fft.fftfreq(nof_chan, chan_width)*1e9
+            for pol in range(nof_pols):   #probably unecessary - could do with some matrix mult stuff
+                #some binary logic
+                stream_idx = int(str(tune)+str(pol),2)
+                residual_delays[stream_idx] = -1.0 * tlags[max_idxs[stream_idx]]
+                gain_from_residual_delay = np.exp(1j*np.pi*observation_frequencies[tune,:]*residual_delays[stream_idx])
```

Comment:
I have tried `2pi` here as well as the `pi`. Neither worked.

Reply:
If I understand the variables, pretty sure it should be `2pi` (since `2pi * 1j * freq * delay` is phase; i.e., when delay = 1 period, phase is unchanged)

Reply:
And by freq do we mean `np.arange(1024)` or the actual frequency values?

Reply:
Tried that and that didn't work nvm...

Reply:
Yes I think Jack is right, the observation frequencies (well at least the total BW) needs to come into the delay calculation.  Currently I think we may just be implicitly relying on total BW being 1 GHz which will make delays naturally come out in ns, that should be cleaned up.

Reply:
If you posted these as not-images they would be much easier to plot :)

Reply:
I'm not sure, but I think these might be consistent with the different delays obtained via fit (high-precision) and ifft (2 sample precision)

![image](https://user-images.githubusercontent.com/1402492/216667424-8b6a2812-d671-409d-acd4-fb985b61b401.png)

Reply:
Yeah, subtracting off the linear fit of the ifft (just gradient * chan_number):